packages/create-remix/templates/fly-stack/app/routes/join.tsx

```diff
@@ -1,76 +1,138 @@
+  errors: {
+    email?: string;
+    password?: string;
+  };
+}
+
+export const action: ActionFunction = async ({ request }) => {
   const formData = await request.formData();
   const email = formData.get("email");
   const password = formData.get("password");
+  const returnTo = formData.get("returnTo");
+
+  if (typeof email !== "string" || email.length === 0) {
+    return json<ActionData>(
+      { errors: { email: "Email is required" } },
+      { status: 400 }
+    );
+  }
 
-  invariant(typeof email === "string", "email must be a string");
-  invariant(typeof password === "string", "password must be a string");
+  if (typeof password !== "string" || password.length === 0) {
+    return json<ActionData>(
+      { errors: { password: "Password is required" } },
+      { status: 400 }
+    );
+  }
 
   const user = await createUser(email, password);
 
   if (!user) {
-    return redirect("/login");
+    return json<ActionData>(
+      { errors: { email: "Invalid email or password" } },
+      { status: 400 }
+    );
   }
 
-  return createUserSession(request, user.id, "/");
+  return createUserSession(
+    request,
+    user.id,
+    typeof returnTo === "string" ? returnTo : "/"
+  );
 };
 
-const meta: MetaFunction = () => ({
-  title: "Join"
-});
+export const meta: MetaFunction = () => {
+  return {
+    title: "Join"
+  };
+};
 
-function JoinPage() {
+export default function JoinPage() {
   const [searchParams] = useSearchParams();
   const returnTo = searchParams.get("redirectTo") ?? undefined;
+  const actionData = useActionData<ActionData>();
 
   return (
-    <div>
-      <div>
-        <h2>Join</h2>
-      </div>
-
-      <Form method="post">
+    <>
+      <h1>Join</h1>
+      <Form
+        method="post"
+        style={{ display: "flex", flexDirection: "column", gap: 8 }}
+      >
         <input type="hidden" name="redirectTo" value={returnTo} />
         <label>
           <span>Email address</span>
-          <input name="email" type="email" autoComplete="email" />
+          <input
+            name="email"
+            type="email"
+            autoComplete="email"
+            aria-invalid={actionData?.errors?.email ? true : undefined}
+            aria-errormessage={
+              actionData?.errors.email ? "email-error" : undefined
+            }
+          />
+          {actionData?.errors?.email && (
+            <Alert style={{ color: "red", paddingTop: 4 }} id="email-error">
+              {actionData.errors.email}
+            </Alert>
+          )}
         </label>
 
         <label>
           <span>Password</span>
-          <input name="password" type="password" autoComplete="new-password" />
+          <input
+            name="password"
+            type="password"
+            autoComplete="new-password"
+            aria-invalid={actionData?.errors?.password ? true : undefined}
+            aria-errormessage={
+              actionData?.errors.password ? "password-error" : undefined
+            }
```

Comment:
@ryanflorence and @chaance, do we need this logic or could it simply be

```suggestion
            aria-errormessage="password-error"
```

I've done the ternary before but I'm not sure whether it's necessary. If there's no `password-error` element is that ok to have it here anyway?

Reply:
"The aria-errormessage should only be used when the value of an object is not valid;" https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-errormessage